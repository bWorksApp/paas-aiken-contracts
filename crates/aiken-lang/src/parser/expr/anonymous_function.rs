use crate::{
    ast,
    expr::{FnStyle, UntypedExpr},
    parser::{annotation, error::ParseError, pattern, token::Token},
};
use chumsky::prelude::*;

pub fn parser(
    sequence: Recursive<'_, Token, UntypedExpr, ParseError>,
) -> impl Parser<Token, UntypedExpr, Error = ParseError> + '_ {
    just(Token::Fn)
        .ignore_then(
            params()
                .separated_by(just(Token::Comma))
                .allow_trailing()
                .delimited_by(just(Token::LeftParen), just(Token::RightParen)),
        )
        .then(just(Token::RArrow).ignore_then(annotation()).or_not())
        .then(sequence.delimited_by(just(Token::LeftBrace), just(Token::RightBrace)))
        .map_with_span(
            |((arguments, return_annotation), body), span| UntypedExpr::Fn {
                arguments,
                body: Box::new(body),
                location: span,
                fn_style: FnStyle::Plain,
                return_annotation,
            },
        )
}

pub fn params() -> impl Parser<Token, ast::UntypedArg, Error = ParseError> {
    // TODO: return a better error when a label is provided `UnexpectedLabel`
    choice((
        select! {Token::DiscardName {name} => name}.map_with_span(|name, span| {
            ast::ArgBy::ByName(ast::ArgName::Discarded {
                label: name.clone(),
                name,
                location: span,
            })
        }),
        select! {Token::Name {name} => name}.map_with_span(|name, span| {
            ast::ArgBy::ByName(ast::ArgName::Named {
                label: name.clone(),
                name,
                location: span,
            })
        }),
        pattern().map(ast::ArgBy::ByPattern),
    ))
    .then(just(Token::Colon).ignore_then(annotation()).or_not())
    .map_with_span(|(by, annotation), span| ast::UntypedArg {
        is_validator_param: false,
        location: span,
        annotation,
        doc: None,
        by,
    })
}

#[cfg(test)]
mod tests {
    use crate::assert_expr;

    #[test]
    fn anonymous_function_basic() {
        assert_expr!(r#"fn (a: Int) -> Int { a + 1 }"#);
    }

    #[test]
    fn anonymous_function_by_pattern_no_annotation() {
        assert_expr!(r#"fn (Foo { my_field }) { my_field * 2 }"#);
    }

    #[test]
    fn anonymous_function_by_pattern_with_annotation() {
        assert_expr!(r#"fn (Foo { my_field } : Foo) { my_field * 2 }"#);
    }

    #[test]
    fn anonymous_function_by_pattern_with_alias() {
        assert_expr!(r#"fn (Foo { my_field, .. } as x) { my_field * my_other_field }"#);
    }
}
