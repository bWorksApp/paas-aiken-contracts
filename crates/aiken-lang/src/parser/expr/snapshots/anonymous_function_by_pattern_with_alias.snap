---
source: crates/aiken-lang/src/parser/expr/anonymous_function.rs
description: "Code:\n\nfn (Foo { my_field, .. } as x) { my_field * my_other_field }"
---
Fn {
    location: 0..60,
    fn_style: Plain,
    arguments: [
        UntypedArg {
            by: ByPattern(
                Assign {
                    name: "x",
                    location: 4..29,
                    pattern: Constructor {
                        is_record: true,
                        location: 4..24,
                        name: "Foo",
                        arguments: [
                            CallArg {
                                label: Some(
                                    "my_field",
                                ),
                                location: 10..18,
                                value: Var {
                                    location: 10..18,
                                    name: "my_field",
                                },
                            },
                        ],
                        module: None,
                        constructor: (),
                        spread_location: Some(
                            20..22,
                        ),
                        tipo: (),
                    },
                },
            ),
            location: 4..29,
            annotation: None,
            doc: None,
            is_validator_param: false,
        },
    ],
    body: BinOp {
        location: 33..58,
        name: MultInt,
        left: Var {
            location: 33..41,
            name: "my_field",
        },
        right: Var {
            location: 44..58,
            name: "my_other_field",
        },
    },
    return_annotation: None,
}
